dask/utils.py

```diff
@@ -1257,3 +1259,64 @@ def ndimlist(seq):
         return 1
     else:
         return 1 + ndimlist(seq[0])
+
+
+class _HashIdWrapper(object):
```

Comment:
I'm curious, why do we need this?

Reply:
Hashing a large tuple takes time proportional to the length of the tuple (about 2ms for 10^6 elements on my machine), which would somewhat negate the benefits of caching (although it is still faster than np.cumsum, mostly because converting a tuple to an ndarray is surprisingly slow). So I'm trading off poorer cache usage (because tuples with the same value but different identity require separate cache slots) to get higher performance lookups.

Reply:
If we didn't add the HashIdWrapper, and hashed tuples directly, how much slower would that make your workloads?

When I look at the difference I see about a 20x difference

```python
In [1]: t = tuple(range(1000000))

In [2]: d = {}

In [3]: %time d[t] = 1
CPU times: user 2.54 ms, sys: 0 ns, total: 2.54 ms
Wall time: 2.55 ms

In [4]: %timeit d[t]
2.31 ms ± 151 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [5]: import numpy as n

In [6]: import numpy as np

In [7]: %timeit np.cumsum(t)
43.1 ms ± 477 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

I'm hoping that this is fast enough, and that something else becomes the bottleneck instead, especially given your statement here:

> A quick benchmark shows that the result for indexing a single value from
a large array is similar to that from #4909, i.e., around 10x faster for
constructing the graph.

Reply:
I'll take a look, but my goal is to make slicing M chunks out of an N-chunk array take O(M + log N) rather than O(N). Are you objecting to the wrapper because it reduces cache re-use or just because it adds complexity?

Reply:
Just because of complexity.  It's not a big deal, we can keep it (it's well isolated), I just thought I'd ask in case it wasn't necessary in practice.

Reply:
Here's some benchmarking. It's somewhat pathological (an enormous number of very small chunks) so I don't expect the differences to be as significant in real use, but it shows that the wrapper does improve performance.

```python
#!/usr/bin/env python

import time

from katdal.lazy_indexer import dask_getitem
import dask.array as da
import numpy as np


a = da.ones((1000000, 100), chunks=(1, 100))
step = 100
start = time.monotonic()
for i in range(0, a.shape[0], step):
    dask_getitem(a, np.s_[i : i + step]).compute()
stop = time.monotonic()
print(stop - start)
```

- master: 1301 seconds
- branch: 278 seconds
- branch without `_HashIdWrapper`: 339 seconds

Reply:
It improves things, but not by a large amount.  My guess is that if we didn't use `_HashIdWrapper` that you wouldn't notice enough to add it.

I'm happy either way, with a slight preference for simpler/slower code.  It sounds like you prefer HashIdWrapper though, given that you seem to be making a few changes these days I'm happy to defer to your preference here.